commands/run.sh

```diff
@@ -135,18 +135,21 @@ fi
 set +e
 
 (
-  # Reset bash to the default IFS with no glob expanding and no failing on error
+  # Reset bash to the default IFS
   unset IFS
-  set -f
 
-  # The eval statements below are used to allow $BUILDKITE_COMMAND to be interpolated correctly
-  # When paired with -f we ensure that it word splits correctly, e.g bash -c "pwd" should split
-  # into [bash, -c, "pwd"]. Eval ends up the simplest way to do this, and when paired with the
-  # set -f above we ensure globs aren't expanded (imagine a command like `cat *`, which bash would
-  # helpfully expand prior to passing it to docker-compose)
+  # Because we receive $BUILDKITE_COMMAND as a single string, it needs to be tokenized as a shell
+  # would to respect things like quoted strings. Previously we used eval to do this with some
+  # careful control around preventing glob expansion, but this ends up being difficult to get just
+  # right and also very risky. The new implementation uses xargs and printf to tokenize the string,
+  # which is portable and much harder to shoot ourselves in the foot 🎉
+
+  while IFS= read -rd '' token; do
+    [[ -n "$token" ]] && run_params+=("$token")
+  done < <(xargs printf '%s\0' <<< "$BUILDKITE_COMMAND")
```

Comment:
I don’t fully grok how this works. What’s the `/0`?

Reply:
The first interesting bit is that xargs tokenizes quoted strings, which was news to me:

```
# echo "blah 'a long string'" | xargs -n1 echo token
token blah
token a long string
```

Then the next bit of magic is getting those tokens into bash so that we can loop over them. Because quoted strings can contain any normal delimiters like newlines or semi-colons or whatever, we use null bytes `\0` to delimit our tokens so that we can loop over them. But because bash is c-based, it's slightly allergic to null bytes, because they tend to mean the end of strings, so to produce the null bytes we use `printf` to output the tokens with a \0 at the end.

Then to read them, we set `IFS` (the field separator that is normally used for word-splitting) to empty and then use the `-d ''` param in the `read` built-in, which will split on those null bytes! 

So the end result is to tokenize the `$BUILDKITE_COMMAND` into arguments, respecting single and double quotes and then iterate over them and put them in an array as strings. 🎉

Reply:
This is an excellent story that belongs in the code!

> Because quoted strings can contain any normal delimiters like newlines or semi-colons or whatever, we use null bytes \0 to delimit our tokens so that we can loop over them. 

Is this because it's impossible for people the contents of the args themselves to contain null bytes, so we can use them to split?

Reply:
> Is this because it's impossible for people the contents of the args themselves to contain null bytes, so we can use them to split?

Yup correct, because if there was a \0 in there it would cause the string to terminate.